### ISSUE TITLE
Iteration of ICA data representation

### ISSUE THREAD

**Issue by ColmMassey** [SolidarityEconomyAssociation/ica-project#3]
In schema:
 - [x] Change spelling of Fishery > Fishing
 - [x] Change spelling of Utility ->Utilities
 - [x] Change spelling of Wholesale and Retail ->Wholesale and retail trade
 - [x] Consumer co-operative to Consumer/User coop 
 - [ ] ~~Create another Property which we should call organisationalTypology which also uses the organisational-structure VES.~~
- [x] Create a new VES called base-membership-type with 6 values
  -  Consumer/Users
  -  Producers
  -  Workers
  -  Multi-stakeholders
  -  Residents
  -  Others
- [x] Create a new Property called baseMembershipType which uses base-membership-type VES

When importing:
 - [x] If Cooperative is in the Structure Type, put it in organisationalStructure
 - [x] Populate the new baseMembershipType from Typology column in data

In dialog:
 - [x] Display Economic Activity rather than just Activity
 - [x] Display Structure Type rather than Organisational Structure 
 - [x] Display baseMembershipType as Typology
 - [ ] ~~Write the plural of any organisationalTypology. Just at an 's'!~~
 - [x] Replace last 's' with ' coops'
 - [x] If there is not an organisationTypology present, display 'Others'. 
 - [x] If it has a Qualiifier Coop Promoter/Supporter, display it under Structure Type

![Kenya](https://user-images.githubusercontent.com/1162642/94046967-4ec4de00-fdc9-11ea-98d2-9885b5598cb2.png)

**Comment by dtmakm27** [comment]
> When importing:
 If Cooperative is in the Structure Type, put it in organisationalStructure
 Populate the new organisationalTypology from Typology column in data if not listed as Other



don't quite understand this, could you try briefly explaining it in another way? 


organisationalStructure -> old one that we were using
organisationalTypology -> new property
correct?

**Comment by dtmakm27** [comment]
@ColmMassey

**Comment by ColmMassey** [comment]
I predicted that we would need a chat about this. Please suggest some times for a Zoom.

**Comment by dtmakm27** [comment]
When importing:

 If Cooperative is in the Structure Type, put it in organisationalStructure
 Populate the new organisationalTypology from Typology column in data if not listed as Other




We decided to change this slightly: 
instead of a field for organisationalTypology we decided to create a new VES called BaseMembershipType. 
This field will have 5 values:

- Consumer/User coops
- Others
- Producer coops
- Worker coops
- Multi-stakeholder coops


(value will be determined from the Typology field)


organisationalStructure will be used as usual (apart from the case where Structure Type is Cooperative, in that case the value that organisationalStructure will receive is only Cooperative instead of Producer/Worker/Multistakeholder/etc.. Cooperative ) 

@ColmMassey is this correct?

**Comment by ColmMassey** [comment]
Pretty much. I've updated the instructions at the top of the Issue.

**Comment by dtmakm27** [comment]
Hey, i implemented everything and tested it out. 
Could you check out to make sure @ColmMassey 
https://dev.ica.solidarityeconomy.coop/

**Comment by wu-lee** [comment]
I see there's a new standard CSV schema field with ID `:base_membership_type`. I noticed this when integrating my code - I have a merge conflict. This has some consequences I want to share, and raises a couple of questions.

The reason I noticed: I've de-duplicated the Oxford and Newbridge converter code, so that I don't have to duplicate it a third time when creating Mersey Green.

Earlier I saw that a new field `:qualifiers` was added to the standard schema. It wasn't being handled in Oxford/Newbridge. This is possible because of the older conversion code uses a class written by Matt called `SeOpenData::CSV::Standard::V1`, and the newer code use `SeOpenData::CSV::Schema` (shared code) and `SeOpenData::CSV::Schemas` (specific cases) to define the fields.

I've fixed it - for now I've added support for the `:qualifiers` field for Oxford/Newbridge by just inserting a blank field. @dtmakm27 - I hope that's the right thing to do?

I've also marked `SeOpenData::CSV::Standard::V1` deprecated, and it now delegates to `SeOpenData::CSV::Schemas::Latest` to get the fields. This was the direct cause of the merge conflict. But main thing is the class should work as it did before, but without the possibility of adding fields in one case but not the other.

Going forward in general: we want to lose this `V1` class eventually, and use `Schemas`.

Now: the new field `:base_membership_type`. I'll need to add that to the definition in `Schemas` too, and I think again all I can do is insert a blank field for non-ICA projects. Right @dtmakm27 ?

**Comment by ColmMassey** [comment]
> I see there's a new standard CSV schema field with ID `:base_membership_type`. I noticed this when integrating my code - I have a merge conflict. This has some consequences I want to share, and raises a couple of questions.
> 
> The reason I noticed: I've de-duplicated the Oxford and Newbridge converter code, so that I don't have to duplicate it a third time when creating Mersey Green.
> 
> Earlier I saw that a new field `:qualifiers` was added to the standard schema. It wasn't being handled in Oxford/Newbridge. This is possible because of the older conversion code uses a class written by Matt called `SeOpenData::CSV::Standard::V1`, and the newer code use `SeOpenData::CSV::Schema` (shared code) and `SeOpenData::CSV::Schemas` (specific cases) to define the fields.
> 
> I've fixed it - for now I've added support for the `:qualifiers` field for Oxford/Newbridge by just inserting a blank field. @dtmakm27 - I hope that's the right thing to do?
> 
> I've also marked `SeOpenData::CSV::Standard::V1` deprecated, and it now delegates to `SeOpenData::CSV::Schemas::Latest` to get the fields. This was the direct cause of the merge conflict. But main thing is the class should work as it did before, but without the possibility of adding fields in one case but not the other.
> 
> Going forward in general: we want to lose this `V1` class eventually, and use `Schemas`.
> 
> Now: the new field `:base_membership_type`. I'll need to add that to the definition in `Schemas` too, and I think again all I can do is insert a blank field for non-ICA projects. Right @dtmakm27 ?

This deserves its own Issue I think and a longer discussion.

**Comment by ColmMassey** [comment]
> Change spelling of Fishery > Fishing
> Change spelling of Utility ->Utilities
> Change spelling of Wholesale and Retail ->Wholesale and retail trade
They seem to be changed in data, but not in dialog?

>If it has a Qualiifier Coop Promoter/Supporter, display it under Structure Type
Still write Structure Type as when it is actually from Structure Type.

**Comment by ColmMassey** [comment]
Make 'Services' in ICA280 - all lowercase. 
Other Services - Other services

**Comment by dtmakm27** [comment]
@wu-lee 
> I've fixed it - for now I've added support for the :qualifiers field for Oxford/Newbridge by just inserting a blank field. @dtmakm27 - I hope that's the right thing to do?


Yup



> Now: the new field :base_membership_type. I'll need to add that to the definition in Schemas too, and I think again all I can do is insert a blank field for non-ICA projects. Right @dtmakm27 ?



can we actually not do that and default to empty if it doesn't exist?

**Comment by dtmakm27** [comment]
>Change spelling of Fishery > Fishing
Change spelling of Utility ->Utilities
Change spelling of Wholesale and Retail ->Wholesale and retail trade
They seem to be changed in data, but not in dialog?



@ColmMassey  aph, yes, totally forgot we have them separated fixed this


> If it has a Qualiifier Coop Promoter/Supporter, display it under Structure Type
Still write Structure Type as when it is actually from Structure Type. 


Could you send a picture of what you mean here, I don't quite understand what you mean?



> Make 'Services' in ICA280 - all lowercase.
Other Services -> Other services

is this for the vocabulary & data or just the popup?

**Comment by wu-lee** [comment]
> can we actually not do that and default to empty if it doesn't exist?

Can you elaborate? For instance, what does "empty" mean here?  I'm currently setting both `:quantifiers` and `:base_membership_type` unconditionally to an empty string. Do you mean that, or nil, or something else?

**Comment by dtmakm27** [comment]
> Can you elaborate? For instance, what does "empty" mean here? I'm currently setting both :quantifiers and :base_membership_type unconditionally to an empty string. Do you mean that, or nil, or something else?



ah sorry @wu-lee  i meant that instead of specifying the field as empty in every data edition (i'm referencing this): 
> I've fixed it - for now I've added support for the :qualifiers field for Oxford/Newbridge by just inserting a blank field



can we just update the converter to default to empty if the field isn't found

**Comment by ColmMassey** [comment]
> This deserves its own Issue I think and a longer discussion.

**Comment by ColmMassey** [comment]
> @ColmMassey aph, yes, totally forgot we have them separated fixed this

' don't understand this sentence

**Comment by ColmMassey** [comment]
> Could you send a picture of what you mean here, I don't quite understand what you mean?
> 
> > Make 'Services' in ICA280 - all lowercase.
> > Other Services -> Other services
> 
> is this for the vocabulary & data or just the popup?

I think we need a voice call.my only appointment today is between 2 and 3 if you here time today for a quick call.

**Comment by wu-lee** [comment]
> can we just update the converter to default to empty if the field isn't found?
> (i am probably missing something here, but do you get the idea? :D )

@dtmakm27 - Possibly you should give me a call when you're available to check nothing's being missed?

In this specific case, it can't do this because there is no information corresponding do `:qualifiers` in the Oxford, Newbridge (and now Mersey) input data - so far as I know. But since this field doesn't seem to be used by these projects I plugged the gap by setting it to `''`, otherwise quite rightly there's an error saying this hasn't been defined. 

Ditto `:base_membership_type`.

The other projects are all unique snowflakes and may do something different. However, if the new `SeOpenData::CSV::Schema` class is used, it should at least raise an error if this field defined in the output schema but is not set. If it's using the older conversion code,I'm not sure but it may not.

Being able to check these things was one of the central motivations for writing the `Schema` class.

(Note also that catching such discrepancies when the standard output schema changes does *also* require specifying the output schema as `SeOpenData::CSV::Schemas::Latest`, and not using specific version like `SeOpenData::CSV::Schemas::Version[1], which was a problem in the "LimeSurvey" case (Oxford, Newbridge, Mersey) until I fixed it recently.)

**Comment by ColmMassey** [comment]
@dtmakm27 
> If it has a Qualiifier Coop Promoter/Supporter, display it under Structure Type

In this case, we are printing Coop Promoter/Supporter in the  Structure Type field, but we don't say 
 Structure Type: Coop Promoter/Supporter

**Comment by ColmMassey** [comment]
In Economic Activities/Sectors (ICA - L2) VES - ICA 270

Education / Health / Social Work -> 
Education / health / social work

**Comment by ColmMassey** [comment]
These last few tweaks are trivial. Can you quickly change them and push to dev.ica first think Monday?

**Comment by dtmakm27** [comment]
> In Economic Activities/Sectors (ICA - L2) VES - ICA 270

shall i change it only on the map view or in the VES as well?

**Comment by ColmMassey** [comment]
> shall i change it only on the map view or in the VES as well?

It is already lowercase in the VES, so only needed in the ICA dialog.

**Comment by dtmakm27** [comment]
Yes, I changed it everywhere (VES, ruby code, and map), so that we're consistent 

![image](https://user-images.githubusercontent.com/34026440/96607468-4c827080-12f0-11eb-9bc7-dd68132f9204.png)

**Comment by dtmakm27** [comment]
https://dev.ica.solidarityeconomy.coop

**Comment by ColmMassey** [comment]
I'll wait until the caching is rolled out to ICA to test.

**Comment by ColmMassey** [comment]
Somehow the Typology field has now disappeared! 

https://dev.ica.solidarityeconomy.coop/?showDatasetsPanel=False&htmlTitle=ICA%20Members&initialBounds=62,-107,-37,174&maxZoomOnOne=10

**Comment by ColmMassey** [comment]
And please push to the production server when ready.
Both sea-maps are 0.1.57 on producation and dev, so it's just differences with databases. Prod has more failed geocoding. Probably not using same source file that dev used. Manual changes were made to csv used for dev.

**Comment by ColmMassey** [comment]
> Somehow the Typology field has now disappeared!

Can you take a look at this @wu-lee ?

**Comment by wu-lee** [comment]
I'm not really sure what the Typology field is or how to recover it, even looking back over this issue, seems like the raison d'etre was discussed on other channels. However, my hopeful hypothesis is that there was code Dean hasn't committed back to Github, which he deployed, but was rolled back when I redeployed subsequently. Trying to get in touch with @dtmakm27 to check this.

**Comment by dtmakm27** [comment]
> Somehow the Typology field has now disappeared!

https://dev.ica.solidarityeconomy.coop/?showDatasetsPanel=False&htmlTitle=ICA%20Members&initialBounds=62,-107,-37,174&maxZoomOnOne=10




Not entirely sure how that happened but it was indeed missing, now it's fixed, @ColmMassey 
I'm assuming i've missed pushing to dev but if it occurs again i'll explore

**Comment by dtmakm27** [comment]
> And please push to the production server when ready.
Will update both the db and the sea-map for prod

**Comment by ColmMassey** [comment]
How are you testing this? As I still can't see it.

**Comment by dtmakm27** [comment]
try in incognito, seems to have cached the website on your end (js and html)

**Comment by ColmMassey** [comment]
We can't get users to use incognito, so I don't want to test using incognito.

**Comment by ColmMassey** [comment]
We need to solve this somehow.

**Comment by dtmakm27** [comment]
> Have I found new caching problem, or are we just playing cache Wakamole? @wu-lee @dtmakm27 ?


new one that we have not discussed much: it's the browser (on your machine) basically caching the resources (javascript, css and html) so it doesn't have to retrieve them all the time. I have seen fixes for this and I think we have an issue for it somewhere, I can take a look and try to fix this next if you'd like? @ColmMassey

**Comment by dtmakm27** [comment]
TLDR: 
New caching Issue, unrelated to this one

**Comment by ColmMassey** [comment]
Moving cacheing discussion to https://github.com/SolidarityEconomyAssociation/technology-and-infrastructure/issues/51

**Comment by wu-lee** [comment]
To be fair, browser caching issues are (and have been) problems since the dawn of the web, and not something arising from this project specifically.

Usually the ETAG/Last-changed headers set automatically by the server (same ones we use for checking data modification) allow the browser to handle this automatically at the level of the web server/browser interaction. When it doesn't work the browser's debug console (network tab) should be checked for [304-Not-Modified][1] statuses first (which means the browser is caching pages because it thinks they've not changed). Typically this shouldn't be a problem for us, because most of the content is static, and the web server reports the date the file was uploaded to the server.  Although sometimes all that seems fine and it seems to be pure stubbornness by the browser.

If it's not that, then the detective work has to follow the trail back to whatever is not changing when it should. HTTP headers need to be checked for cache options, etc. (also visible in the browser debug console), and the problem needs to be replicated (which is often the hardest part).

[1]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/304

**Comment by ColmMassey** [comment]
@dtmakm27 How did you test the latest update of ica sea-map and data? Did you look at the Kenya Bank example?
It looks like something has been rolled back again? When I look ot both dev.ica and prod.ica on a private browser in the dialog for the example above, Co-operative Bank of Kenya Ltd (CBK) the "Typology:" text is no longer present and the Typology itself is wrong. The dialog says _Other_ when in the data it is _Consumer/Users_

https://prod.data.solidarityeconomy.coop/ica/17.html

**Comment by dtmakm27** [comment]
Yes, just saw that, thanks for prompting me. Accidently used old query for getting the linked data.
It's using the latest map-app and data for both now:
http://dev.ica.solidarityeconomy.coop/
http://prod.ica.solidarityeconomy.coop/

**Comment by ColmMassey** [comment]
Can you update dev. now to use the new tiles?
